### PR TITLE
Add access information to SymbolInformation

### PR DESCRIPTION
**Pull request by tanishiking**
https://github.com/lampepfl/dotty/issues/12963

This PR adds `access` information to `SymbolInformation`.  I confirm the diff is legit just by thoroughly check the diff in `metac.expect`.

(but, it might be better to have a script to verify the Scala2 and Scala3 will extract the same SemanticDB for the same program, so we can use the scala2's output as a test oracle 🤔 )(as long as the program compiles both in Scala2 and Scala3)

### CODE REVIEW

File: tests/semanticdb/metac.expect [hunk 1]
```diff
@@ -12,12 +12,12 @@ Occurrences => 19 entries
...
 example/Access#`<init>`(). => primary ctor <init>
-example/Access#m1(). => method m1
-example/Access#m2(). => method m2
-example/Access#m3(). => method m3
-example/Access#m4(). => method m4
-example/Access#m5(). => method m5
-example/Access#m6(). => method m6
+example/Access#m1(). => method private[this] m1
+example/Access#m2(). => method private[this] m2
+example/Access#m3(). => method protected[example/Access#] m3
```
Comment (bishabosha) [inline tests/semanticdb/metac.expect:17]: if we look in `tests/semanticdb/expect/Access.scala` this should be `private` qualified not `protected` qualified

File: compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala [hunk 2]
```diff
@@ -450,13 +451,31 @@ class ExtractSemanticDB extends Phase:
...
+        case _ =>
+          if (sym.privateWithin == NoSymbol)
+            if (sym.isAllOf(PrivateLocal)) PrivateThisAccess()
+            else if (sym.is(Private)) PrivateAccess()
+            else if (sym.isAllOf(ProtectedLocal)) ProtectedThisAccess()
+            else if (sym.is(Protected)) ProtectedAccess()
+            else PublicAccess()
+          else
+            val ssym = symbolName(sym.privateWithin)
+            if (sym.isPrivate) PrivateWithinAccess(ssym)
```
Comment (bishabosha) [inline compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala:467]: I think here that a private qualified symbol never gets the private flag, you should instead check if it is protected. See the other comment about the `metac.expect` output for `tests/semanticdb/expect/Access.scala`

File: tests/semanticdb/metac.expect [hunk 3]
```diff
@@ -3808,13 +3808,13 @@ flags/p/package.Y# => final class Y
 flags/p/package.Y#`<init>`(). => primary ctor <init>
 flags/p/package.Z# => sealed trait Z
 flags/p/package.Z#`<init>`(). => primary ctor <init>
-flags/p/package.`y_=`(). => var method y_=
+flags/p/package.`y_=`(). => var method protected y_=
 flags/p/package.`y_=`().(x$1) => param x$1
 flags/p/package.m(). => macro m
 flags/p/package.m().[TT] => typeparam TT
-flags/p/package.x. => lazy val method x
+flags/p/package.x. => lazy val method protected[flags/p/] x
```
Comment (bishabosha) [inline tests/semanticdb/metac.expect:3815]: I'm not sure this is right, in scala 2 semanticdb the access is printed as `private`, and the definition in  `tests/semanticdb/expect/semanticdb-Flags.scala` is `private lazy val x = 1`

**Review by bishabosha — CHANGES_REQUESTED** [review]
Thank you for taking this on!
There seems to be an issue with private-qualified/protected-qualified that does not match my expectation

File: tests/semanticdb/metac.expect (hunk 3 above)
Comment (bishabosha) [reply tests/semanticdb/metac.expect:3815]: to debug you can try using this script I made, save it to `compiler/test/dotty/tools/DottySymbolStealer.scala`

If I run it with
```
scala3-compiler-bootstrapped/Test/runMain dotty.tools.printSymFlags "package object flags { private lazy val x = 1 }" "x"
```

I get the output 
```
val x => lazy <touched> private[module class flags]
```

```scala
package dotty.tools

import dotc.ast.tpd
import dotc.core.Names._
import dotc.ast.tpd._
import dotc.core.Contexts.Context
import dotc.core.Symbols.Symbol
import dotc.core.Decorators._
import dotc.core.Types.Type

@main def printSymFlags(source: String, termRefs: String*) = {
  val (ctx, syms) = DottySymbolStealer.stealSymbol(source, termRefs*)
  given Context = ctx
  syms.map(sym =>
    val priv = sym.privateWithin
    s"$sym => ${sym.flags.flagStrings(if priv.exists then s"$priv" else "").mkString(" ")}"
  ).foreach(println)
}

object DottySymbolStealer extends DottyTest {
  def stealSymbol(source: String, termRefs: String*): (Context, List[Symbol]) = {
    var scontext : Context = null
    var tp: List[Symbol] = null
    checkCompile("typer", source) {
      (tree, context) =>
        given Context = context
        val remaining = scala.collection.mutable.Set.from(termRefs)
        val findValDef: (List[ValDef], tpd.Tree) => List[ValDef] =
          (acc , tree) =>  tree match {
            case t: ValDef =>
              remaining.find(t.name.toString == _) match
                case Some(name) =>
                  remaining -= name
                  t :: acc
                case _ =>
                  acc
            case _ => acc
          }
        val d = new DeepFolder[List[ValDef]](findValDef).foldOver(Nil, tree)
        tp = d.map(_.symbol).reverse
        scontext = context
    }
    (scontext, tp)
  }
}
```

Comment (bishabosha) [reply tests/semanticdb/metac.expect:3815]: so using scala 2 repl it appears that private values in the package object do not get a `privateWithin` symbol, but I think its ok to keep it in scala 3

File: compiler/src/dotty/tools/dotc/semanticdb/Tools.scala [hunk 4]
```diff
@@ -85,6 +85,20 @@ object Tools:
...
+  private def accessString(access: Access): String =
+    access match
+      case Access.Empty => ""
+      case _: PublicAccess => ""
+      case _: PrivateAccess => "private "
+      case _: ProtectedAccess => "protected "
+      case _: PrivateThisAccess => "private[this] "
+      case _: ProtectedThisAccess => "protected[this] "
+      case PrivateWithinAccess(ssym) =>
+        s"private[${ssym}] "
```
[the remark this answers is not in the document]
Comment (tanishiking) [reply compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:97]: Yeah, the situation will be better with https://github.com/lampepfl/dotty/pull/12885 but I was just a bit lazy to duplicate the pretty printer to this PR... Can I go with this and prettify it in the future?

[the remark this answers is not in the document]
Comment (bishabosha) [reply compiler/src/dotty/tools/dotc/semanticdb/Tools.scala:97]: yep, this is fine for now

File: compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala (hunk 2 above)
Comment (tanishiking) [reply compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala]: You're right, fixed in https://github.com/lampepfl/dotty/pull/12964/commits/fb6b679db92e46e8517f6590be4771d6f331ec34 👍

File: tests/semanticdb/metac.expect (hunk 1 above)
Comment (tanishiking) [reply tests/semanticdb/metac.expect]: Fixed by https://github.com/lampepfl/dotty/pull/12964#discussion_r661940556

File: compiler/src/dotty/tools/dotc/semanticdb/Tools.scala [hunk 5]
```diff
@@ -119,7 +133,7 @@ object Tools:
       case TRAIT => sb.append("trait ")
       case INTERFACE => sb.append("interface ")
       case UNKNOWN_KIND | Unrecognized(_) => sb.append("unknown ")
-    sb.append(info.displayName).nl
+    sb.append(s"${accessString(info.access)}${info.displayName}").nl
```
[the remark this answers is not in the document]
Comment (tanishiking) [reply compiler/src/dotty/tools/dotc/semanticdb/Tools.scala]: done https://github.com/lampepfl/dotty/pull/12964/commits/338e6d216b8271e19fad551c46d213d2e16ebf51

**Review by bishabosha — APPROVED** [review]
This looks great to me, thank you

**Pull request merged**